src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java

```diff
@@ -132,24 +134,26 @@
   /**
    * The requested server version.
    */
-  private ExchangeVersion requestedServerVersion = ExchangeVersion.Exchange2010_SP2;
+  private final ExchangeVersion requestedServerVersion;
 
   /**
    * The server info.
    */
   private ExchangeServerInfo serverInfo;
 
-  private Map<String, String> httpHeaders = new HashMap<String, String>();
-
-  private Map<String, String> httpResponseHeaders = new HashMap<String, String>();
+  private Map<String, String> httpHeaders = new ConcurrentHashMap<String, String>();
```

Comment:
I had a typo that was critical here:
"A single instance of ExchangeService can be used by a single server."  I meant a single instance is used by a single thread.  i.e. ExchangeService is not thread safe, but it is OK to have 3 threads using 3 different instances of ExchangeService connecting to same or difference servers in parallel.

It seems that you want to have a single instance of ExchangeService used by multiple threads in parallel.  What is the motivation for that.  That adds complexity and will have weird cases if the state can depend on the request (e.g. if headers may need to change based on request for example).

> I believe we do need to use ConcurrentHashMap here. I believe that ExchangeService should represent
> an Exchange server, and as such, it can be used by multiple threads concurrently.

If we go with the model that I'm suggesting, then that we don't need this protection against concurrency.